### PR TITLE
Enhancement what MODULES get included in the reovery system

### PR DESCRIPTION
**Pull request by jsmeix**
Implemented support for additional meta values like
MODULES=( 'none' )
MODULES=( 'all' )
MODULES=( 'loaded' )
in build/GNU/Linux/400_copy_modules.sh
what MODULES get included in the recovery system
see https://github.com/rear/rear/issues/1202
and https://github.com/rear/rear/issues/1355

### CODE REVIEW

**Comment by jsmeix** [comment]
Do not yet merge it.
This is only a very first attempt mainly FYI
so that you could have a very first look.
For now I only tested that the current default behaviour
did not change for me (on SLES12), nothing else is tested yet.

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh [hunk 1]
```diff
@@ -1,28 +1,129 @@
...
+    # The special user setting MODULES=( 'all' ) enforces that
+    # all files from the /lib/modules/* directories
+    # get included in the rescue/recovery system
+    # regardless of what is set in EXCLUDE_MODULES.
+    # Test all MODULES array members to make the 'all' functionality work for
+    # MODULES array contents like MODULES=( 'moduleX' 'all' 'moduleY' ):
+    if IsInArray "all" "${MODULES[@]}" ; then
+        LogPrint "Copying all kernel modules in /lib/modules/* (MODULES contains 'all')"
+        # The '--parents' is needed to get the '/lib/' directory in the copy:
+        if ! cp $verbose -t $ROOTFS_DIR -a --parents /lib/modules 1>&2 ; then
```
Comment (schlomo) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: How about restricting that to the kernel version that is used? Why should we copy all the modules from all the other potentially installed kernels?

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh [hunk 2]
```diff
@@ -1,28 +1,129 @@
...
+    LogPrint "Omit copying kernel modules (MODULES contains 'none')"
+    return
+fi
+
+# Artificial 'for' clause that is run only once
+# to be able to 'continue' with the code after it
+# when the kernel modules have been copied into the rescue/recovery system
+# (the 'for' loop is run only once so that 'continue' is the same as 'break')
+# which avoids dowdy looking code with deeply nested 'if...else' conditions:
+for dummy in "once" ; do
```
Comment (schlomo) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: IMHO you could achieve the same with this code
```shell
if test "none" = "$MODULES" ; then
  LogPrint "Omit copying kernel modules (MODULES contains 'none')"
  return
elif IsInArray "all" "${MODULES[@]}" ; then
  ... copy all modules ...
elif IsInArray "loaded" "${MODULES[@]}" ; then
  ... copy loaded modules ...
else
  for module in "${MODULES[@]}" ; do
    ... copy individual modules ...
  done
fi
```

and IMHO this reads much easier than the fake `for` loop.

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh [hunk 3]
```diff
@@ -1,28 +1,129 @@
...
-# copy modules & depmod
-LogPrint "Copying kernel modules"
-ModulesCopyTo "$ROOTFS_DIR" "${MODFILES[@]}" >&8
-StopIfError "Could not copy kernel modules"
+done
 
-depmod -avb "$ROOTFS_DIR" "$KERNEL_VERSION" >&8
-StopIfError "Could not configure modules with depmod"
+# Generate modules.dep and map files in all existing $ROOTFS_DIR/lib/modules/* directories
+# (because of the MODULES=( 'all' ) case several such directories could exist):
```
Comment (schlomo) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: Please give more context for taking kernel modules for more than one kernel version. IMHO we will never need this as long as we copy only a single kernel version into the rescue system.

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh [hunk 4]
```diff
@@ -1,28 +1,129 @@
...
+        for module_file in $module_files ; do
+            if ! cp $verbose -t $ROOTFS_DIR -L --preserve=all --parents $module_file 1>&2 ; then
+                Error "Failed to copy $module_file"
+            fi
+        done
+    done
 
-# copy modules & depmod
-LogPrint "Copying kernel modules"
-ModulesCopyTo "$ROOTFS_DIR" "${MODFILES[@]}" >&8
```
Comment (schlomo) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: Is ModulesCopyTo now obsolete? Then please remove it. However, I would suggest to extract such things into functions if it can be tested independently of the other script or used elsewhere.

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh [hunk 5]
```diff
@@ -1,28 +1,129 @@
...
-# copy kernel modules for Relax-and-Recover
+# Copy kernel modules to the rescue/recovery system.
+
 #
 # This file is part of Relax-and-Recover, licensed under the GNU General
 # Public License. Refer to the included COPYING for full text of license.
 
-Log "Collecting modules for kernel version $KERNEL_VERSION"
-MODFILES=(
-$( ResolveModules "$KERNEL_VERSION" "${MODULES[@]}" "${MODULES_LOAD[@]}"  )
```
Comment (schlomo) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: Is ResolveModules now obsolete? Then please remove it. However, I would suggest to extract such things into functions if it can be tested independently of the other script or used elsewhere.

**Comment by ubeaut** [comment]
Hi.
I am new to this git stuff and I don't know if this is the correct place for this comment.
I have noticed that if I create a rescue disk on a headless system (no keyboard or screen) and then I have to do a restore the usb keyboard doesn't work. So if I need to enter anything on the screen I can't.
I got around this problem by using a computer without a usb keyboard.
Once I restored the system I plugged a USB keyboard into the restored system and then created a rescue disk and then I tested that rescue disk and the usb keyboard worked. I could enter things via keyboard.
So my question is how can I insert the modules for a usb keyboard into the rescue disk?
Maybe have the usb modules for keyboard loaded by default?
Other than that problem the restore worked well.
Thanks

**Comment by jsmeix** [comment]
@ubeaut
please use a separated GitHub issue
to report things with USB keyboards.
In general when reporting issues see
https://raw.githubusercontent.com/rear/rear/master/.github/ISSUE_TEMPLATE.md
In particular regarding USB keyboards have a look
at the comments in
https://github.com/rear/rear/pull/1244
Do you use current ReaR GitHub master code?

**Comment by jsmeix** [comment]
@schlomo 

I agree that MODULES=( 'all' ) should mean
all modules that match $KERNEL_VERSION
because no other kernel exists in the recovery system.

Regarding the ResolveModules and ModulesCopyTo
functions: For my very first attempt I do all directly
in the script. Later I will see to what extent it helps
to extract things into separated functions. Then
I may adapt linux-functions.sh appropriately.

Regarding the artificial 'for' loop:
Depending on the particular case I prefer it
over lengthy global 'if elif else fi' or 'case' constructs
which add unnecessary levels of nesting and indentation.
The artificial 'for' loop is my general way to make an
arbitrary code block wherefrom I can jump forward
and out at any place.
It is my way to implement 'goto' as it is normally used
in C to jump straight forward usually to some kind of
"finishing" or "cleanup" code at the bottom, cf.
http://stackoverflow.com/questions/245742/examples-of-good-gotos-in-c-or-c

**Comment by jsmeix** [comment]
Now modules in EXCLUDE_MODULES are generically removed
regardless if the module name or the module filename is
specified in EXCLUDE_MODULES, cf.
https://github.com/rear/rear/pull/1244#issuecomment-286718122

Now EXCLUDE_MODULES are removed in any case,
also for MODULES=( 'all' ) and MODULES=( 'loaded' )
so that the user has the final power.

**Comment by schlomo** [comment]
@jsmeix, I think that goto doesn't lead to clean code so that I would
really prefer an if construct here. If the blocks are too long for your
taste then this is rather a sign that you should extract the block into a
function instead.

**Comment by jsmeix** [comment]
@schlomo
I think we need to agree that we disagree here
because from my point of view that 'goto' usage
leads to cleaner code (of course "from my point of view"
because otherwise I would not implement it this way).
Perhaps there is a more obvious way in bash
how to "jump out and straight forward"?
I think it is a matter of taste who understands
some kind of coding patterns better than others.

**Comment by jsmeix** [comment]
I noticed that I have in "rear -d -D mkrescue" log
things like (excerpts):
<pre>
+ source /root/rear/usr/share/rear/rescue/GNU/Linux/260_storage_drivers.sh
...
++ IsInArray Module rbd xen-blkback ... nosy find: ''\''/lib/modules/4.4.21-69-default/kernel/drivers/ide'\'':' No such file or directory pata_sch ...
</pre>

**Comment by jsmeix** [comment]
According to
<pre>
git log -p --follow usr/share/rear/rescue/GNU/Linux/230_storage_and_network_modules.sh
</pre>
it seems
rescue/GNU/Linux/230_storage_and_network_modules.sh
was falsely fixed via
https://github.com/rear/rear/commit/d6d1ff253a244dde4dbf5e0b5676150d028f1cf7
that reads
<pre>
dropping find error messages
</pre>
but implements it by adding '2>&1' that results
<pre>
..._DRIVERS=( $( find ... 2>&1 | sed ... ) )
</pre>
which adds the find error messages to the module arrays.
I really appreciate the commment in that commit that tells what
it intent is so that I can now easily fix it (without blind guessing
what might have been actually meant with strange looking code).

**Comment by jsmeix** [comment]
FYI: This is still work in progress and
under somewhat heavy development
but I think I make good progress towards
more reliable, predicatable, and configurable
kernel modules inclusion in the recovery system.

**Comment by jsmeix** [comment]
At first glance the current code seems to work too good for me
(on my SLES12 test system) - so that I believe there must be
at least one evil regression hidden somewhere ;-)

On my SLES12 test system "rear mkbackup/mkrescue"
and "rear recover" work for the default and for
MODULES=( 'all' ) and MODULES=( 'loaded' ).

FYI:
on my SLES12 test system I get
those recovery system ISO sizes
by default: 77M
with MODULES=( 'all' ): 94M
with MODULES=( 'loaded' ): 64M

I made the recovery system ISOs
with gzip default compression
cf. https://github.com/rear/rear/issues/1142
and with FIRMWARE_FILES=( 'No' )
cf. https://github.com/rear/rear/issues/1216

**Comment by jsmeix** [comment]
Now I use those more explicit special MODULES values
MODULES=( 'all_modules' )
MODULES=( 'loaded_modules' )
MODULES=( 'no_modules' )
and I adapted my last comment
https://github.com/rear/rear/pull/1359#issuecomment-300829232
accordingly.

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh [hunk 6]
```diff
@@ -1,28 +1,148 @@
...
+    local module_name=$1
+    local module_filename=""
+    # Older modinfo (e.g. the one in SLES10) does not support '-k'
+    # but that old modinfo returns a zero exit code when called as 'modinfo -k ...'
+    # and shows a 'modinfo: invalid option -- k ...' message on stderr and nothing on stdout
+    # so that we need to check if we got a non-empty module filename:
+    module_filename=$( modinfo -k $KERNEL_VERSION -F filename $module_name 2>/dev/null )
+    # If 'modinfo -k ...' stdout is empty we retry without '-k' regardless why stdout is empty
+    # but then we do not discard stderr so that error messages appear in the log file:
+    test $module_filename || module_filename=$( modinfo -F filename $module_name )
```
Comment (schlomo) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: So you first check on the `$KERNEL_VERSION` kernel and if that fails you check under the currently running kernel and **take that file**? Can you please explain why we should mix kernel versions?

Comment (jsmeix) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: Can you please explain how to deal with
such an broken old modinfo?
Don't the comments explain what goes on?

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh [hunk 7]
```diff
@@ -1,28 +1,148 @@
...
+
+# Local functions that are 'unset' at the end of this script:
+function modinfo_filename () {
+    local module_name=$1
+    local module_filename=""
+    # Older modinfo (e.g. the one in SLES10) does not support '-k'
+    # but that old modinfo returns a zero exit code when called as 'modinfo -k ...'
+    # and shows a 'modinfo: invalid option -- k ...' message on stderr and nothing on stdout
+    # so that we need to check if we got a non-empty module filename:
+    module_filename=$( modinfo -k $KERNEL_VERSION -F filename $module_name 2>/dev/null )
```
Comment (schlomo) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: Why not something like this?
```
modinfo_filename=$( modinfo --set-version $KERNEL_VERSION --filename $module_name 2>&1) || Error "Cannot find find module file for $module_name: $modinfo_filename"
```

And then, if you want to also support old distros, you should explicitly check if `module_filename` contains the error message.

But actually this is a very good example of using feature flags and extracting functionality into functions and not relying on indirect effects. Then you could also have a check if `KERNEL_VERSION` is different from the running kernel and abort with an error message that explains that the user cannot use the -r option on his old Linux.

File: usr/share/rear/rescue/GNU/Linux/230_storage_and_network_modules.sh [hunk 8]
```diff
@@ -1,51 +1,38 @@
...
-		sed -e 's/^\(.*\)\.ko.*/\1/'
-		#  ^^^^- remove the .ko, faster one sed call than many basename calls or shell code
-	)
-)
+Log "Including additional drivers"
+EXTRA_DRIVERS=( $( find_modules_in_dirs /lib/modules/$KERNEL_VERSION/{extra,weak-updates} ) )
+
+# Local functions must be 'unset' because bash does not support 'local function ...'
+# cf. https://unix.stackexchange.com/questions/104755/how-can-i-create-a-local-function-in-my-bashrc
+unset -f find_modules_in_dirs
```
Comment (schlomo) [inline usr/share/rear/rescue/GNU/Linux/230_storage_and_network_modules.sh]: Why not have that and the other function in a suitable `lib/*-functions.sh` file instead?

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh (hunk 7 above)
Comment (jsmeix) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: Didn't you note that the crappy old modinfo does not
return a non-zero exit code so that
modinfo ... || Error
would not work.

If I had just ignored that ols stuff nobody would have cared.
Now I try to somehow deal even with only crap and
all is not right.
I think I simply remove all that overcomplicated attempts
doe work around crappy tools and leave users with crapy
tools alone.

File: usr/share/rear/rescue/GNU/Linux/230_storage_and_network_modules.sh (hunk 8 above)
Comment (jsmeix) [inline usr/share/rear/rescue/GNU/Linux/230_storage_and_network_modules.sh]: Because I like to have functions that are called
in only one single script local in that script.
Why having local things global?

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh (hunk 6 above)
Comment (schlomo) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: Yes, your comment was clear. No need to get offended here.

I have two concerns: 

1. Why do we need to support old distros seemingly forever? IMHO we should be able to draw a line and say that old distros stay on old ReaR versions.

2. Our code should be robust and easy to read. In this example I would rather test which features `modinfo` supports and then do one or the other thing. Trying one and then doing the other if the first didn't give me any result is a very indirect approach that could go wrong in many ways. And it has the additional effect of hiding the actual error message and substituting it with an error message from a modinfo call **without** the kernel version. Which means that on a system where modinfo support -k but you check for a non-existant module you will then check in another kernel version than what you meant.

All that is the result from implicit (or no) error handling and just trying again instead of reacting to the error that modinfo -k gives on newer systems and supporting older systems through an active decision and different code path.

BTW, without quoting your `test` statements will fail if `$module_failname` is empty.

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh (hunk 7 above)
Comment (schlomo) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: Yes, I am indeed against complicating our code for the sake of old or ancient distros.

If we know that old modinfo doesn't work well then IMHO we should prevent the user from shooting himself in the foot. If this is only about using rear with a different kernel version then I would find that a very minor setback. Of course I assume here that old modinfo will throw an error code if the module cannot be found.

Why not simply add a new script to the `prep` stage that checks for this scenario and aborts if a user specifies a different kernel version on old distros?

**Comment by jsmeix** [comment]
My current code works so well for me that I will merge it
so that also others can test it and report if there are issues
or even regressions which I will of course fix.

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh [hunk 9]
```diff
@@ -1,28 +1,153 @@
...
+    # If 'modinfo -k ...' stdout is empty we retry without '-k' regardless why stdout is empty
+    # but then we do not discard stderr so that error messages appear in the log file.
+    # In this case we must additionally ensure that KERNEL_VERSION matches 'uname -r'
+    # otherwise a module file for a wrong kernel version would be found:
+    if ! test $module_filename ; then
+        test "$KERNEL_VERSION" = "$( uname -r )" || Error "modinfo_filename failed because KERNEL_VERSION does not match 'uname -r'"
```
Comment (schlomo) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: Very good. I would prefer this once in the `prep` stage before we start working instead of failing at the first module that we look for.

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh [hunk 10]
```diff
@@ -1,28 +1,153 @@
...
+
+# Local functions that are 'unset' at the end of this script:
+function modinfo_filename () {
+    local module_name=$1
+    local module_filename=""
+    # Older modinfo (e.g. the one in SLES10) does not support '-k'
+    # but that old modinfo returns a zero exit code when called as 'modinfo -k ...'
+    # and shows a 'modinfo: invalid option -- k ...' message on stderr and nothing on stdout
+    # so that we need to check if we got a non-empty module filename:
+    module_filename=$( modinfo -k $KERNEL_VERSION -F filename $module_name 2>/dev/null )
```
Comment (schlomo) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: Why hide STDERR here? It might contain useful infos that should appear in the log. Especially if you bail out on non-matching kernel versions during prep.

**Comment by schlomo** [comment]
Please make sure to also test the negative code flows, e.g. if a module
cannot be found.

**Comment by jsmeix** [comment]
I had tested MODULES=( 'qqqq' ) and got
in "rear -d -D recover" log
<pre>
++ echo -e 'Copying kernel modules'
++ for module in '"${MODULES[@]}"'
++ module=qqqq
++ module=qqqq
++ modinfo qqqq
++ continue
++ for module in '"${MODULES[@]}"'
++ module=nfs
++ module=nfs
++ modinfo nfs
+++ modprobe --ignore-install --set-version 4.4.21-69-default --show-depends nfs
</pre>
i.e. the non-existent module qqqq is silently ignored
which is backward compatible -  i.e. it matches the
behaviour in ResolveModules
<pre>
        for module in "$@"; do
            ...
                # Check if the module actually exists
                if ! modinfo $module &>/dev/null; then
                        continue
                fi
</pre>

**Comment by jsmeix** [comment]
Now I did a more evil test (on my SLES12 system):
<pre>
# modprobe --dry-run --show-depends lp
insmod /lib/modules/4.4.21-69-default/kernel/drivers/parport/parport.ko 
insmod /lib/modules/4.4.21-69-default/kernel/drivers/char/lp.ko 

# mv /lib/modules/4.4.21-69-default/kernel/drivers/parport/parport.ko /lib/modules/4.4.21-69-default/kernel/drivers/parport/away_parport.ko
# mv /lib/modules/4.4.21-69-default/kernel/drivers/parport/parport_pc.ko /lib/modules/4.4.21-69-default/kernel/drivers/parport/away_parport_pc.ko

# modprobe --dry-run --show-depends lp
modprobe: ERROR: ctx=0x22f0010 path=/lib/modules/4.4.21-69-default/kernel/drivers/parport/parport.ko error=No such file or directory
modprobe: ERROR: ctx=0x22f0010 path=/lib/modules/4.4.21-69-default/kernel/drivers/parport/parport.ko error=No such file or directory
insmod /lib/modules/4.4.21-69-default/kernel/drivers/char/lp.ko 
</pre>
I have set in local.conf
MODULES=( 'qqqq' 'lp' )
and did with the old code
<pre>
# usr/sbin/rear -d -D mkrescue
...
Copying binaries and libraries
Copying kernel modules
Omit copying files in /lib*/firmware/ (FIRMWARE_FILES='No')
</pre>
No error about the missing 'parport' modules.

I got same bahavior with my new code
<pre>
# usr/sbin/rear -d -D mkrescue
...
Copying binaries and libraries
Copying kernel modules
Omit copying files in /lib*/firmware/ (FIRMWARE_FILES='No')
</pre>
so that I am sufficiently backward compatible
but I wonder if such errors should better be detected
and at least reported to the user via LogPrint?

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh (hunk 9 above)
Comment (jsmeix) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: As far as I understand it normally KERNEL_VERSION
can be different than the current running kernel version
i.e. as far as I understand it one can runa different kernel
compared to what is used for the recovery system
but I never tested that myself.
But the current code looks as if that should work.
In this particular case and for MODULES=( 'loaded_modules' )
it is different because then I need matching KERNEL_VERSION
with the currently running one.
Therefore those tests are currently at the particular places
where needed.
I don't know if ther is a valis deason why a different kernel
version could be used for the recovery system compared
to the current running kernel.
But perhaps we can simply enforce that the recovery system
always uses what matches the currenly running kernel?

File: usr/share/rear/build/GNU/Linux/400_copy_modules.sh (hunk 10 above)
Comment (jsmeix) [inline usr/share/rear/build/GNU/Linux/400_copy_modules.sh]: I fully agree that in general stderr should be in the log
and I will change that next week.
FYI:
I notice relatively often that stderr is redirected to /dev/null
<pre>
$ find usr/sbin/rear usr/share/rear/ | xargs egrep '2>/dev/null|&>/dev/null' | wc -l
112
</pre>
Perhaps this should be changed basically everywhere
(unless thre is really a good reason to discard stderr).

**Comment by jsmeix** [comment]
@schlomo 
regarding your above question
"Why do we need to support old distros seemingly forever?
 IMHO we should be able to draw a line and say that old
 distros stay on old ReaR versions."

In general see
https://github.com/rear/rear/issues/1173

I do fully agree and SLES10 is already "out of support", see
https://github.com/rear/rear/blob/master/doc/rear-release-notes.txt
at about line 1328:
<pre>
Rear-2.00 dropped officially support for the following
Linux based operating systems:
  o Fedora <22
  o RHEL 3 and 4
  o SLES 9 and 10
  o openSUSE <11
  o Debian <6
  o Ubuntu <12
</pre>

Unfortunately in this particular case I detected the crappy
old modinfo behaviour by "bad luck" while playing around
with the modules stuff also on my SLES10 system (just to
get some experience how things work in a traditional way)
but since I knew that misbehaviour I could not "just ignore" it
so that I tried to make code that should behave reasonably
even for the crappy old modinfo because I do not want
to break SLES10 support knowingly as long as I can keep
it working even on SLES10 (with reasonable effort) and
perhaps also on whatever other "elderly" systems
that may also have that crappy modinfo and
might be even still "supported" by ReaR
(I don't know what there is on RHEL 5 or CentOS 5).

Next time I will not play around on SLES10 ;-)

**Comment by jsmeix** [comment]
With https://github.com/rear/rear/pull/1365 merged
stderr is no longer needlessly redirected to /dev/null
see https://github.com/rear/rear/pull/1359#discussion_r116239195

In particular non-existent modules are now at least reported in the log
cf. https://github.com/rear/rear/pull/1359#issuecomment-301071974

On the other hand useless stderr warnings from 'cp'
are avoided by removing duplicate module files
in the 'modprobe --show-depends' output, cf.
http://blog.schlomo.schapiro.org/2015/04/warning-is-waste-of-my-time.html
